### ISSUE TITLE
[TraceQL] Add trace level scope

### ISSUE THREAD

**Issue by joe-elliott** [grafana/tempo#1989]
**Is your feature request related to a problem? Please describe.**
There is currently no way to write a query on trace level information. This is desirable b/c it will allow more expressive and efficient queries. i.e. searching by trace duration would be far more efficient then searching by span duration

Proposed scope: `trace`
Proposed fields: `duration`, `service.name`, `name`

- duration: max end span time - min start span time
- service.name: `resource.service.name` on the root span
- name: `span.name` on the root span

Example:
```
{ trace.duration > 1s && resource.env = "prod" }
```

**Comment by mdisibio** [comment]
Another benefit is that this closes the gap between the original tag-based search and TraceQL, i.e. it is possible to express all of the tag-based search as a TraceQL query.   Provides opportunity to simplify Tempo internals but also the Grafana UX (thinking of a query builder).

**Comment by joe-elliott** [comment]
So the biggest obstacle to adding this is agreeing on syntax. Proposing the following:

```
{ traceDuration > 1s }  // total trace duration
{ traceName = "foo" }  // matches the root span's name
{ traceServiceName = "bar" } // matches the root span's service name
```

I prefer this syntax b/c it is consistent with all "intrinsic" properties being named directly without scope. Consider the current proto for the [InstrumentationScope](https://github.com/open-telemetry/opentelemetry-proto/blob/7aa439cb0ba78afbd009d83f32fdda6c128e0342/opentelemetry/proto/common/v1/common.proto#L71)) which we [want to support](https://github.com/grafana/tempo/issues/2189).

This requires adding both intrinsics (InstrumentationScope name and version) as well as attributes. In this case we will need to distinguish between intrinsics and attributes at the same scope just like we do with span intrinsics and attributes.
```
{ scopeName = "foo" } // scope level intrinsic
{ scope.name = "foo" } // scope level custom attribute
```
If intrinsics are always named directly without scope then it will help us clearly support InstrumentationScope, Trace and any future scopes.

**Comment by mapno** [comment]
I don't dislike your proposal, but I'd actually prefer to use `trace` scope.

```
{ trace.duration > 1s }  // total trace duration
{ trace.name = "foo" }  // matches the root span's name
{ trace.service.name = "bar" } // matches the root span's service name
```

I don't think the _intrinsics_ arguments holds very well. These attributes are not intrinsic to the trace per se (ie. there isn't a trace-level `name` attribute in the spec), but rather a convention that the values for the root span are taken. I think the scope `trace` can represent that concept better.

Also, I prefer the syntax of dots. It's more consistent with the rest of the language than camelCase.

**Comment by joe-elliott** [comment]
aesthetically i much prefer `trace.duration`. I'm just stuck on the fact that currently every time you invoke a scope `resource.` or `span.` you are always referring to a custom attribute and this breaks that.

What do you suggest we do for the intrinsic scope name vs. an attribute at the scope level with the key "name"?

**Comment by mdisibio** [comment]
I think we need to be careful about distinguishing between root and trace, they are not the same.  I don't think traces technically require a singular root (i.e. unparented) span.  So root duration would be the duration of the root span if present, trace duration would be `max(end) - min(start)`.  

`traceServiceName` could definitely be `root.service.name` which is inline with the original idea of a `root` scope.  If that works we don't need `trace.service.name` (and not technically a thing following the argument above).   

`trace.duration` could match `max(end) - min(start)`, without making a precedent of how `root.duration` should work.

Agree `root.name` doesn't work though because it conflicts with the intrinsic.  `trace.name` doesn't make sense (traces don't have names, it really means root.name).  We put a lot of effort into preventing conflicts between intrinsics and attribute names, we should keep that as a hard design constraint.  There must still be a way to access both of them.

I unfortunately just thought of this conundrum: `root.service.name` probably should be `root.resource.service.name`. 😬

**Comment by bjornbyte** [comment]
I'm not particularly opinionated about the syntax, but how I so far think about queries is that `{ stuff }` refers to particular spans and so `{ .a = "A" } && { .b = "B"}` is traces with a span that has a="A" and b="B", but the whole query there is about the traces, so I kind of expected to do something like `{.service.name="my-service" } && duration > 1s`, where the bit outside of any {}
is applying to the trace as a whole, and inside any {} are referring to particular spans within the trace. I don't know if it breaks the whole paradigm though.  

Since a span does have a traceID, it is also reasonable to do something like `{ trace.duration >1s }` as more of a span referencing the trace it belongs to type of approach.  I'm not sure if it's more confusing our useful though to do something like `{ .a="A" && trace.duration >1s } && { b="B" && trace.name="foo"}`., but that's something to consider here.

**Comment by bjornbyte** [comment]
that looks sensible to me.  Not that since the root span is the same across the whole trace your example query would be equivalent to 
`{ root_span.resource.service.name="gateway" && root_span.resource.deployment.environment="produciton" && span.messaging.system="rabbitmq" } && trace.duration > 10s` (i.e.  can use separate span targets or the same), but I think that's fine, probably unavoidable even.

**Comment by adirmatzkin** [comment]
Coming back to the "root span" discussion - we should probably keep it out of this issue.
Just to mention - I recently opened a new issue about a [parent span scope](https://github.com/grafana/tempo/issues/2492).
I believe the root span scope is unnecessary if a parent span scope would become a feature - it is contained in it (something like `{ parent=nil && more_logic_here }`) plus it has more to it...

**Comment by joe-elliott** [comment]
Adding three trace level intrinsics [here](https://github.com/grafana/tempo/pull/2503) that will cover a lot of the needs discussed in this thread.

A true "root" scope will be more difficult, but we'll keep this thread open to track it.

**Comment by bjornbyte** [comment]
> Adding three trace level intrinsics [here](https://github.com/grafana/tempo/pull/2503) that will cover a lot of the needs discussed in this thread.
> 

I agree, the addition of `traceDuration`, `rootServiceName` and `rootName` covers most of this.  The main thing that I think is lacking now is that those aren't very discoverable since they don't pop up in the auto-complete it Temo as far as I can tell (unlike say `name`).  Is that an easy thing to fix?

**Comment by joe-elliott** [comment]
Oof. Thank you for the reminder! I normally file an issue with Grafana when I add features like this. Linked grafana issue is above :point_up:

**Comment by stoewer** [comment]
TraceQL now supports the trace scope for intrinsics and allows queries such as `{ trace:id="1b629cd6065ca0295203ffbfa918374a" }` or `{ trace:duration>500ms }`

@joe-elliott I think we can close this issue. What do you think?

**Comment by joe-elliott** [comment]
Agreed! Nice catch.
